### ISSUE TITLE
Ultra seqfaults with -p > 2781

### ISSUE THREAD

**Issue by marade** [TravisWheelerLab/ULTRA#66]
If it's a memory issue, it would be nice if it told me that. If it's a code issue, it would be nice if it supported higher values. Thanks for writing this software.

**Comment by traviswheeler** [comment]
We've used ULTRA with larger values of -p and 2781, so there's something specific happening. Segfault means that it's somehow escape our error handling, which likely means that the "something specific" is something we haven't prepared for.

If you'd like us to try to figure it out, please provide us with a bit more information. Things like this are a good start:
- What are the characteristics of the dataset that you're analyzing with ULTRA? (size of the genome? anything interesting about it, like number of sequences? is it a publicly-available genome, and if yes, is there a link?).  
- What was the entire command that you ran (this helps us understand if there may be some interplay with other flags)
- What are the characteristics of the machine you're using? (What OS? How much RAM?)
- Does ULTRA work for other data sets that you've tested? Or for other choices of -p? (I get the impression that -p 2781 works, but -p 2782 does not; is that right?

Basically: more information to us ==> better chance that we can figure out what's going on.

**Comment by marade** [comment]
I was hoping it was known issue... In any case I always appreciate a prompt response from the PI. This is for five genomes, 6-7Mb each. All of them seqfault at -p 2782 and not at -p 2781. Sorry, they aren't publicly available, but I decided to test on one that is...

https://www.ncbi.nlm.nih.gov/nuccore/110645304

...and yes, same error profile. (I downloaded the complete nucleotide sequence from that site, getting a sequence.fasta file that appears to be proper genome nucleotide sequence.) The only other flag used is -o. The OS is Ubuntu 22.04.4 on an AWS EC2 c7i-4xlarge instance.

**Comment by DanielOlson** [comment]
Running`ultra -p 2782 Pseudomonas_aeruginosa.fa` works fine for me. Running `ultra -p 2782 --mem` reveals that the run is expected to take around 32 GB, and so the AWS EC2 c7i-4xlarge is likely running out of memory. I just pushed an update (1.0.2b) that will reduce the default memory usage for large models and also always print the expected memory usage before performing analysis. In 1.0.2b you should use a bit over 17 GB when running `./ultra -p 2782`

Large period repeat finding is computationally challenging. Two setting adjustments I have found useful when running ULTRA with a large maximum repeat period: 
- `-i 2 -d 2`. This reduces the number of indels ULTRA is able to detect, and in doing so reduces memory/runtime for large period repeats by a factor of 5.
- `--win_size <at least 10X the repeat period>`: This reduces the size of ULTRA's DP matrix, allowing for (potentially) large reductions in memory usage. Decreasing the window size does cause a slight increase in runtime, and may also slightly lower the annotation quality. I generally avoid this unless I am working with very very large models.

I'm closing this issue, but please open a new issue if problems persist.

**Comment by marade** [comment]
Thanks! I appreciate the help with this issue.
